docs/dev_guide/code_coverage_guide.md

```diff
@@ -0,0 +1,105 @@
+coverage is below the threshold. The threshold, targets at the branch coverage, is defined in the `build.gradle` of each 
+submodule independently.
+
+### Example commands
+```Bash
+# Template
+./gradlew :<module name>:<submodule name>:jacocoTestCoverageVerification 
+
+./gradlew :clients:venice-push-job:jacocoTestCoverageVerification 
+```
+
+The report will be located at `<module name>/<submodule name>/build/reports/jacoco/test/index.html`.
+
+
+### Debugging notes
+
+#### Run commands against module doesn't generate the report
+
+Due to the current project setup, running the commands, e.g. `./gradlew :clients:jacocoTestCoverageVerification`, doesn't
+execute the unit tests thus no jacoco report will be generated. Please be sure to run the commands against the submodule.
+
+## New-commit coverage verification
+
+### Description
+This will check the overall code coverage at the commit-level.
+DiffCoverage, which is an extension of Jacoco, gathers the diff files by comparing the local branch and remote upstream 
+main branch, and leverage the Jacoco report, to re-generate a new report only for these newly added lines/files. Similarly,
+it performs the coverage verification and fail the build if the coverage is below the threshold. The threshold is defined 
+at 60% for branch coverage.
+
+### Example commands
+```Bash
+# Template
+./gradlew :<module name>:<submodule name>:jacocoTestCoverageVerification diffCoverage
+
+./gradlew :clients:venice-push-job:jacocoTestCoverageVerification diffCoverage 
+```
+
+The report will be located at `<module name>/build/reports/jacoco/diffCoverage/html/index.html`.
+
+### Debugging notes
+
+#### Integ tests are added but DiffCoverage doesn't identify the code coverage of it
+Though Integ tests is strongly encouraged, both Jacoco and DiffCoverage only work with unit tests so please consider writing 
+unit tests.
+
+#### Unit tests are added but the DiffCoverage report doesn't reflect my changes
+
+There are two possible reasons.
+1. **Jacoco report isn't up-to-date**. DiffCoverage report relies on the Jacoco Report to reflect the correct coverage. 
+If the Jacoco report isn't up-to-date, for example, executing the wrong command `./gradlew :clients:venice-push-job:diffCoverage` which
+misses the step of re-running unit tests and updating the Jacoco report, can cause this issue. 
+   1. Please be sure to run test and Jacoco report first.
+2. **Units tests are placed in a different module**. Jacoco can only analyze and generate reports based off the current 
+submodule. So, if you write new source codes in `submodule A` and `submodule B`, and you only have written unit tests in
+`submodule B` which cover changes in `submodule A`, this cannot be recognized by Jacoco thus the DiffCoverage doesn't think
+there's coverage too.
+   1. Please move/re-organize some unit tests to the right submodule such that the coverage can be detected and reported.
```

Comment:
IDK if this is considered best practice or not but it doesn't feel quite right... (I've been known to disagree with best practices so there's that...)

Is it possible for Jacoco to evaluate coverage of/by dependent modules as well?

I wonder if the module structure is serving us well if we need to play these shenanigans... Almost tempting to fuse everything into a single module if that's how code coverage is assessed holistically (not ideal, but still seems like it's the proposed incentive...).

Reply:
> Is it possible for Jacoco to evaluate coverage of/by dependent modules as well?

I don't think Jacoco can be that smart but alternatively, we can merge Jacoco's `.exec` files together so there will be only one place to see the Jacoco report and diffCoverage report. But this also means, we will have to execute the entire test suite to find out what we miss in coverage report, and this can be time-consuming.

> IDK if this is considered best practice or not but it doesn't feel quite right 
...
I wonder if the module structure is serving us well if we need to play these shenanigans

I think the module structure is great and really helps me to navigate codes. I also think it's great to add unit tests in the corresponding module, 1) it's easy to find the corresponding unit test file when you make changes, 2) be responsible for other modules. Imaging if we separate one module as a dedicated lib and vend it to external code bases, it's definitely reasonable to have unit tests in its own module.